src/tir/schedule/primitive/loop_transformation.cc

```diff
@@ -0,0 +1,420 @@
+    }
+  }
+
+  /*! \brief The predicate to be appended */
+  const PrimExpr& to_append_;
+  /*! \brief Whether to add predicate */
+  bool add_predicate_;
+};
+
+/*! \brief Substitute vars and collect the reuse mapping of opaque blocks */
+class SubstituteVarAndCollectOpaqueBlock : public StmtExprMutator {
+ public:
+  explicit SubstituteVarAndCollectOpaqueBlock(std::function<Optional<PrimExpr>(const Var&)> vmap,
+                                              Map<Block, Block>* opaque_blocks)
+      : vmap_(vmap), opaque_blocks_(opaque_blocks) {}
+
+ private:
+  PrimExpr VisitExpr_(const VarNode* op) final {
+    Var var = GetRef<Var>(op);
+    if (Optional<PrimExpr> ret = vmap_(var)) {
+      return ret.value();
+    } else {
+      return std::move(var);
+    }
+  }
+
+  Stmt VisitStmt_(const BlockRealizeNode* op) final {
+    Stmt res = StmtMutator::VisitStmt_(op);
+    if (op->block->iter_vars.empty()) {
+      const BlockRealizeNode* realize = TVM_TYPE_AS(realize, res, BlockRealizeNode);
+      opaque_blocks_->Set(op->block, realize->block);
+    }
+    return res;
+  }
+
+  /*! \brief The substitute function */
+  std::function<Optional<PrimExpr>(const Var&)> vmap_;
+  /*! \brief The reuse mapping of opaque blocks */
+  Map<Block, Block>* opaque_blocks_;
+};
+
+/*! \brief Simplify the binding of block realize and update the opaque block reuse mapping */
+class IterMapSimplifyBlockBinding : public StmtExprMutator {
+ public:
+  explicit IterMapSimplifyBlockBinding(const Map<Var, Range>& loop_map,
+                                       Map<Block, Block>* opaque_blocks)
+      : opaque_blocks_(opaque_blocks), loop_var2extent_(std::move(loop_map)) {}
+
+  static For SimplifyBindings(const Stmt& stmt, const Array<StmtSRef>& loop_srefs,
+                              Map<Block, Block>* opaque_blocks) {
+    Map<Var, Range> loop_var2extent;
+    for (const StmtSRef& sref : loop_srefs) {
+      const ForNode* loop = TVM_SREF_TO_FOR(loop, sref);
+      loop_var2extent.Set(loop->loop_var, Range::FromMinExtent(loop->min, loop->extent));
+    }
+    return Downcast<For>(IterMapSimplifyBlockBinding(loop_var2extent, opaque_blocks)(stmt));
+  }
+
+ private:
+  Stmt VisitStmt_(const ForNode* op) final {
+    loop_var2extent_.Set(op->loop_var, Range::FromMinExtent(op->min, op->extent));
+    Stmt res = StmtMutator::VisitStmt_(op);
+    loop_var2extent_.erase(op->loop_var);
+    return res;
+  }
+
+  Stmt VisitStmt_(const BlockRealizeNode* op) final {
+    // skip opaque block and update mapping
+    if (op->iter_values.empty()) {
+      Stmt res = StmtMutator::VisitStmt_(op);
+      const BlockRealizeNode* realize = res.as<BlockRealizeNode>();
+      MapNode* mutable_map = opaque_blocks_->CopyOnWrite();
+      for (const std::pair<Block, Block>& entry : *opaque_blocks_) {
+        if (entry.second.same_as(op->block)) {
+          mutable_map->at(entry.first) = realize->block;
+          break;
+        }
+      }
```

Comment:
1. It is possible that `op` has become a dangling pointer, so please do not use it.
2. We don't need to iterate the map to find out a single element, use `MapNode::find` instead. If it is certain that the element exists, we can do it in a single line:

```C++
      map->at(block) = realize->block;
```

Reply:
2 is not correct, because I compare `op->block` to `entry.second` instead of `entry.first`, so I can't do it in a single line.

Reply:
Sorry I misread. Thanks for the explanation!